File: fibonacci.py
# реализация через классы и кучи
class FibonacciNode:
    def __init__(self, key):
        self.key = key
        self.degree = 0
        self.mark = False
        self.parent = None
        self.child = None
        self.left = self
        self.right = self

class FibonacciHeap:
    def __init__(self):
        self.min_node = None
        self.total_nodes = 0
        selfRoots = []

    def insert(self, key):
        # реализация вставки узла
      pass

    def extract_min(self):
        # реализация извлечения минимума
      pass

File: fibonacci.cpp
#include <iostream>
#include <vector>
#include <climits>

// Структура узла
struct Node {
    int key;
    int degree;
    Node* parent;
    Node* child;
    Node* left;
    Node* right;
    bool mark;

    Node(int k) : key(k), degree(0), parent(nullptr), child(nullptr), mark(false) {
        left = right = this;
    }
};

// Класс кучи Фибоначчи
class FibonacciHeap {
private:
    Node* minNode;
    int totalNodes;

public:
    FibonacciHeap() : minNode(nullptr), totalNodes(0) {}

    // Вставка узла
    void insert(int key) {/* ... */}

    // Извлечение минимального узла
    int extractMin() {/* ... */}
}
